### PR TITLE
[WIP] v1alpha2 and conversion

### PR DESCRIPTION
**Pull request by knelasevero**
Closes https://github.com/kubernetes-sigs/descheduler/issues/926

Opening the PR for early eyes on it since I got tests to pass. Still have to change some things, add more stuff to the conversion, and write more tests (also need to work on defaults, validation, etc). But please feel free to comment on anything that you might feel could be improved.

In the issue we said we should have every strategy in v1alpha1  converted into a separate v1alpha2 profile, but here I prefered converting into a single profile. In the old API it was like we had a single profile, I don't see how having them separate helps with individual configuration. Specially with the filter and Evictor, that can all be used in a single profile.

I also did not get rid of `pkg/descheduler/strategy_migration.go` yet. I am not sure how we plan to do that. I am also not sure how we are planning to get rid of all the type assertions and type castings.

### CODE REVIEW

**Comment by knelasevero** [comment]
I did not touch e2e yet, they will probably fail here in these first runs

**Comment by damemi** [comment]
> In the issue we said we should have every strategy in v1alpha1 converted into a separate v1alpha2 profile, but here I prefered converting into a single profile

I agree with this (sorry I missed the discussion in the issue). It's the simplest conversion imo, unless there is a technical reason why separate profiles would be better I think a user who is still on v1alpha1 isn't going to expect profile behavior

File: pkg/descheduler/descheduler.go [hunk 1]
```diff
@@ -342,90 +304,44 @@ func RunDeschedulerStrategies(ctx context.Context, rs *options.DeschedulerServer
...
+				!rs.DisableMetrics,
+				eventRecorder,
+			)
+
+			// for now running sequentially
+			// TODO(knelasevero): implement profiles in parallel
+			enabledDeschedulePlugins := profile.Plugins.Deschedule.Enabled
+			enabledBalancePlugins := profile.Plugins.Balance.Enabled
+
+			for _, pluginName := range enabledDeschedulePlugins {
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/descheduler/descheduler.go:328]: @knelasevero we have two options here:
- go through all the profiles and extension points and invoke them (as done here)
- keep wiring the profiles/plugins in strategy_migration.go and implement invocation of the extension points through framework runAllDeschedule/runAllBalance EPs later.

**Comment by ingvagabund** [comment]
> In the issue we said we should have every strategy in v1alpha1 converted into a separate v1alpha2 profile, but here I prefered converting into a single profile. In the old API it was like we had a single profile, I don't see how having them separate helps with individual configuration. Specially with the filter and Evictor, that can all be used in a single profile.

Defaulting of NodeFit alongside https://github.com/kubernetes-sigs/descheduler/issues/640#issuecomment-1216506965.

> I also did not get rid of pkg/descheduler/strategy_migration.go yet. I am not sure how we plan to do that. I am also not sure how we are planning to get rid of all the type assertions and type castings.

This will be relevant once we introduce the framework runAllDeschedule/runAllBalance EPs with addtional handlers and eviction per node/namespace limits.

File: pkg/descheduler/descheduler.go [hunk 2]
```diff
@@ -436,6 +352,72 @@ func RunDeschedulerStrategies(ctx context.Context, rs *options.DeschedulerServer
...
+	return nil
+} 
+
+func getSortersFiltersAndEvictorConfigFromProfile(profile v1alpha2.Profile) (v1alpha2.PluginConfig, v1alpha2.PluginConfig, v1alpha2.PluginConfig) {
+	var evictorConfig v1alpha2.PluginConfig
+	var filterConfig v1alpha2.PluginConfig
+	var preEvictionFilterConfig v1alpha2.PluginConfig
+
+	ecivtorPluginName :=  profile.Plugins.Evict.Enabled[0]
+	filterPluginName :=  profile.Plugins.Filter.Enabled[0]
```
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go]: But I don't see how them both would be enabled :thinking:

File: pkg/descheduler/descheduler.go [hunk 3]
```diff
@@ -436,6 +352,72 @@ func RunDeschedulerStrategies(ctx context.Context, rs *options.DeschedulerServer
...
+	preEvictionFilterPluginName :=  profile.Plugins.PreEvictionFilter.Enabled[0]
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/descheduler/descheduler.go]: You can have the `DefaultEvictor` plugin with one `evict` and one `preEvictionFilter` extension point. And another custom plugin with just one `preEvictionFilter` extension point. Having both plugins enabled and chaining `preEvictionFilter` extension points of both plugins.

File: pkg/descheduler/descheduler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go:328]: https://github.com/kubernetes-sigs/descheduler/issues/979

Opened an issue, will continue with this PR focusing on the conversion and then come back tho this

File: pkg/descheduler/descheduler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go]: Are we changing the handle evictorImpl to receive multiple evictor plugins with alternative filters then? Is this supposed to happen now in this PR?

File: pkg/descheduler/policyconfig.go [hunk 4]
```diff
@@ -39,17 +50,298 @@ func LoadPolicyConfig(policyConfigFile string) (*api.DeschedulerPolicy, error) {
...
+func validateDeschedulerConfiguration(in v1alpha2.DeschedulerPolicy) error {
+	return nil
+}
+
+func strategyToProfiles(strategies v1alpha1.StrategyList) (*[]v1alpha2.Profile, error) {
+	var profiles []v1alpha2.Profile
+	for name, strategy := range strategies {
+		switch name {
+		case "RemoveDuplicates":
+			removeduplicatesArgs := convertRemoveDuplicatesArgs(strategy.Params)
```
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/policyconfig.go:171]: What I am doing here is the opposite of in-lining, no? Are you giving a +1 thumbs up, or asking me to inline the bodies of those functions back here? :sweat_smile:

[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/policyconfig.go:171]: Oh, I guess you mean that it is good I am not using the type-unsafe approach with direct assigning

File: pkg/descheduler/descheduler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go]: I sent some code trying to implement this. Feels a bit weird, but let me know what you think

File: pkg/descheduler/descheduler.go [hunk 5]
```diff
@@ -436,6 +352,72 @@ func RunDeschedulerStrategies(ctx context.Context, rs *options.DeschedulerServer
...
+
+	evictorFilter, _ := defaultevictor.New(
+		filterArgs,
+		&handleImpl{
+			clientSet:                 clientSet,
+			getPodsAssignedToNodeFunc: getPodsAssignedToNode,
+			sharedInformerFactory:     sharedInformerFactory,
+		},
+	)
+	handle := &handleImpl{
```
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go]: heavy_check_mark moved it up

File: pkg/descheduler/descheduler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (knelasevero) [reply pkg/descheduler/descheduler.go]: see https://github.com/kubernetes-sigs/descheduler/pull/974#discussion_r994659888

**Comment by knelasevero** [comment]
> Any arg validation should be performed during the conversion of loading from the policy file.

I will work on tests + validation + defaulting now. But feel free to send more comments!

**Comment by knelasevero** [comment]
Ok, I guess I went pretty far from what is the standard kubernetes decoding/converting/validating/defaulting. 

@spike-liu since you are registering the defaulting and validation in a local schema of the plugins, I shouldn't have to manually call them, right? I wonder how can I add those stuff so the [UniversalDecoder](https://github.com/kubernetes-sigs/descheduler/pull/974/files#r991274322) it already does conversion/validation/defaulting :thinking: 

I was having a look over https://github.com/kubernetes/kubernetes/blob/83415e5c9e6e59a3d60a148160490560af2178a1/pkg/scheduler/apis/config/v1/conversion.go#L31 and I tried a few things, but I could not get it to work here

**Comment by spike-liu** [comment]
> Ok, I guess I went pretty far from what is the standard kubernetes decoding/converting/validating/defaulting.
> 
> @spike-liu since you are registering the defaulting and validation in a local schema of the plugins, I shouldn't have to manually call them, right? I wonder how can I add those stuff so we can make the [UniversalDecoder](https://github.com/kubernetes-sigs/descheduler/pull/974/files#r991274322) already do conversion/validation/defaulting 🤔
> 
> I was having a look over https://github.com/kubernetes/kubernetes/blob/83415e5c9e6e59a3d60a148160490560af2178a1/pkg/scheduler/apis/config/v1/conversion.go#L31 and I tried a few things, but I could not get it to work here

Sorry, @knelasevero. I am new to this project and my implementation is fully based on @damemi's design. Hence I am really not sure about how to call them.

**Comment by knelasevero** [comment]
@ingvagabund what was the reasoning behind moving plugin args into their own package? I noticed that for KS they still keep it under the config package (kind of analog to our api package) and there it seems simpler to work with schemes/decoding/conversions :thinking: 

Right now the only thing missing is that when decoding things are coming as raw data and not getting turned into actual objects, and I think plugin args are not being registered correctly still in schemes and I can't figure out why :sob: 

Everything else seems to work, currently

**Comment by knelasevero** [comment]
Yeah, I am trying something like that: https://github.com/kubernetes-sigs/descheduler/blob/c41758ab61ea221b9cd87b0ffe8f67bfe30e14d0/pkg/framework/plugins/removeduplicates/register.go

I am not sure I am doing this correctly though.

**Comment by damemi** [comment]
@knelasevero api registration has always been kind of magic to me... just looking at the diff though we also pass `addKnownTypes` to [`localSchemeBuilder.Register()`](https://github.com/kubernetes-sigs/descheduler/blob/v0.25.1/pkg/api/v1alpha1/register.go#L51) which it doesn't look like you're doing

**Comment by knelasevero** [comment]
Tried that now :( still the decoder loads it as Raw data.

**Comment by ingvagabund** [comment]
https://github.com/ingvagabund/descheduler/commit/e5125167d8d9511ca8b2e2c9fc8df1764d962753 produces the conversion of the args

**Comment by damemi** [comment]
> [ingvagabund@e512516](https://github.com/ingvagabund/descheduler/commit/e5125167d8d9511ca8b2e2c9fc8df1764d962753) produces the conversion of the args

Does this require us to bake-in conversion funcs for each plugin? That can work for now to unblock this, but we'll still need to figure out how to break this requirement, because it won't be feasible for 3rd party plugins

**Comment by knelasevero** [comment]
Yeah, I did not want to do that, so I already implemented the plugin registration already (I know it is a big PR, but I did not want to go with that switch case). Hope it is not a problem. Tests are passing! :raised_hands:

**Comment by knelasevero** [comment]
BTW, thanks for the pointers :partying_face:

File: pkg/framework/plugins/pluginbuilder/pluginbuilder.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+type PluginBuilder = func(args runtime.Object, handle framework.Handle) (framework.Plugin, error)
+
+type Registry = map[string]PluginBuilderAndArgsInstance
+
+func NewRegistry() Registry {
+	return Registry{}
+}
+
+func init() {
+	PluginRegistry = NewRegistry()
```
Comment (knelasevero) [inline pkg/framework/plugins/pluginbuilder/pluginbuilder.go:39]: Storing the plugin registry in memory here

File: pkg/framework/plugins/podlifetime/pod_lifetime.go [hunk 7]
```diff
@@ -131,3 +132,15 @@ func (d *PodLifeTime) Deschedule(ctx context.Context, nodes []*v1.Node) *framewo
 
 	return nil
 }
+
+func init() {
```
Comment (knelasevero) [inline pkg/framework/plugins/podlifetime/pod_lifetime.go:136]: Plugins register themselves like this

File: pkg/api/v1alpha2/conversion.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+			prof.PluginConfig[j].Args = internalArgs
+		}
+	}
+	return nil
+}
+
+func Convert_v1alpha2_PluginConfig_To_api_PluginConfig(in *PluginConfig, out *api.PluginConfig, s conversion.Scope) error {
+	out.Name = in.Name
+	for name, builderAndArgsInstance := range pluginbuilder.PluginRegistry {
+		if in.Name == name {
```
Comment (damemi) [inline pkg/api/v1alpha2/conversion.go:95]: couldn't you just get `pluginbuilder.PluginRegistry[in.Name]` and throw an error if the plugin is not registered?

**Review by damemi — COMMENTED** [review]
@knelasevero awesome, I love that use of the plugin registry. one comment on it but I am +1 for this approach!

File: pkg/api/v1alpha2/conversion.go (hunk 8 above)
Comment (knelasevero) [reply pkg/api/v1alpha2/conversion.go]: You are totally right! :sweat_smile: Been working with those profiles slices all the time and forgot this was a map! Thanks! just pushed

**Comment by ingvagabund** [comment]
@knelasevero before the final review any chance of reducing the number of commits in the PR? Plus, making the changes more iterative so it is easier to review? :)

**Comment by knelasevero** [comment]
> @knelasevero before the final review any chance of reducing the number of commits in the PR? Plus, making the changes more iterative so it is easier to review? :)

That was the plan already. That last commit was sent late in last Friday, so couldn't organise anything.

Obs: Defaulting is still missing here I think, since we are customizing everything and not really letting the scheme handle it. 

While I don't do that, since today is first day of Kubecon and I'm still figuring out badges, internet and stuff, if you want to review the the previous changes comments, that would be helpful (just to see if those changes are properly taken care of). 🙌

**Pull request closed without merging**